src/modular/boxed_residue/mul.rs

```diff
@@ -186,20 +187,22 @@ impl Drop for MontgomeryMultiplier<'_> {
 /// x and y are required to satisfy 0 <= z < 2**(n*_W) and then the result z is guaranteed to
 /// satisfy 0 <= z < 2**(n*_W), but it may not be < m.
 ///
+/// Output is written into the lower (i.e. first) half of `z`.
+///
 /// Note: this was adapted from an implementation in `num-bigint`'s `monty.rs`.
 // TODO(tarcieri): refactor into `reduction.rs`, share impl with `(Dyn)Residue`?
 #[cfg(feature = "alloc")]
-pub(crate) fn montgomery_mul(z: &mut [Word], x: &[Word], y: &[Word], m: &[Word], k: Word) {
+fn montgomery_mul(z: &mut [Word], x: &[Word], y: &[Word], m: &[Word], k: Word) {
     // This code assumes x, y, m are all the same length (required by addMulVVW and the for loop).
     // It also assumes that x, y are already reduced mod m, or else the result will not be properly
     // reduced.
-    let mut c: Word = 0;
     let n = m.len();
+    debug_assert_eq!(z.len(), n * 2);
+    debug_assert_eq!(x.len(), n);
+    debug_assert_eq!(y.len(), n);
+    debug_assert_eq!(m.len(), n);
 
-    let pre_cond = z.len() > n && z[n..].len() == n && x.len() == n && y.len() == n && m.len() == n;
```

Comment:
Why did you remove this, @newpavlov suggested this in zulip, as this removes bounds checks from the code later down the line. Now this will not happen as the `debug_assert`s don't get compiled in release mode

Reply:
It's probably worth to add a comment explaining this together with `z.len() > n && z[n..].len() == n` intentionally being used instead of `z.len() == 2*n`, since the latter prevents compiler from eliminating some bound checks.

Reply:
I tried reverting it and it had no observable change on benchmarks, FWIW